File: backend/app/schemas/runtime.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class RuntimeMetadataResponse(BaseModel):
    app_name: str
    display_name: str
    app_version: str
    current_phase: str | None
    current_phase_number: int | None
    current_phase_name: str | None
    current_phase_summary: str | None
    release_notes_href: str | None
    release_notes_available: bool


class ReleaseNoteResponse(BaseModel):
    phase: str
    display_name: str
    title: str
    summary: str
    published_at: str | None = None
    status: str | None = None
    hero: dict[str, Any] = Field(default_factory=dict)
    highlights: list[dict[str, Any]] = Field(default_factory=list)
    sections: list[dict[str, Any]] = Field(default_factory=list)
    story_refs: list[str] = Field(default_factory=list)
    cta: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/lib/runtime.ts
export type RuntimeMetadata = {
  app_name: string;
  display_name: string;
  app_version: string;
  current_phase: string | null;
  current_phase_number: number | null;
  current_phase_name: string | null;
  current_phase_summary: string | null;
  release_notes_href: string | null;
  release_notes_available: boolean;
};

export type ReleaseNote = {
  phase: string;
  display_name: string;
  title: string;
  summary: string;
  published_at: string | null;
  status: string | null;
  hero: {
    eyebrow?: string;
    headline?: string;
    subheadline?: string;
  };
  highlights: Array<{
    title: string;
    description: string;
  }>;
  sections: Array<{
    title: string;
    items: string[];
  }>;
  story_refs: string[];
  cta: {
    label?: string;
    href?: string;
  };
};

export type AppRoute =
  | { name: "home" }
  | { name: "release-notes"; phaseId?: string };

export function readHashRoute(hash = window.location.hash): AppRoute {
  const normalized = hash.replace(/^#/, "").trim() || "/";
  const parts = normalized.replace(/^\//, "").split("/").filter(Boolean);
  if (parts[0] === "release-notes") {
    return {
      name: "release-notes",
      phaseId: parts[1],
    };
  }
  return { name: "home" };
}
